File: Data Structures & Algorithms/meeting-schedule-ii/submission-2.js
/**
 * Definition of Interval:
 * class Interval {
 *   constructor(start, end) {
 *     this.start = start;
 *     this.end = end;
 *   }
 * }
 */

class Solution {
    /**
     * @param {Interval[]} intervals
     * @returns {number}
     */
    // time - O(n logn), space - O(n)
    minMeetingRooms(intervals) {
        const startTime = []
        const endTime = []
        for (const interval of intervals) {
            startTime.push(interval.start);
            endTime.push(interval.end);
        }
        startTime.sort((a,b) => a-b);
        endTime.sort((a,b) => a-b);
        let active = 0
        let start = 0
        let end = 0
        let maxRoom = 0
        while (start < startTime.length) {
            if (startTime[start] < endTime[end]) {
                active += 1
                start++
                maxRoom = Math.max(maxRoom, active)
            } else {
                end++
                active -= 1
            }
        }
        return maxRoom;
    }
}

File: Data Structures & Algorithms/insert-new-interval/submission-2.py
class Solution:
    # time - O(n), space - O(n)
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        res = []
        start = 0
        end = 1
        for index in range(len(intervals)):
            if intervals[index][end] < newInterval[start]:
                res.append(intervals[index])
            elif newInterval[end] < intervals[index][start]:
                res.append(newInterval)
                res += intervals[index:]
                return res
            else:
                newInterval[start] = min(intervals[index][start], newInterval[start])
                newInterval[end] = max(intervals[index][end], newInterval[end])
        
        res.append(newInterval)
        return res
        

